### PR TITLE
Turn on promise linting and fix issues

### PR DESCRIPTION
**Pull request by nguerrera**
This would have caught https://github.com/Azure/cadl-azure/issues/1297 via static analysis. So I propose turning on the relevant eslint rules to prevent another long debugging session like that. 😁

Everything but playground was pretty clean and some real issues were found and fixed everywhere else without any suppressions. Where issues weren't "real", it still helped to make the code clearer. So I think the value is reasonably high and the noise is reasonably low. 

~~I just opted playground out (for now?). I wasn't familiar enough with the code to be confident changing it and maybe it's too noisy for browser code patterns?~~

*EDIT*: I fixed most of playground now, and only turned off one part of one rule for it since it fires whenever you add an async function as an event listener.

### CODE REVIEW

**Review by bterlson — APPROVED** [review]

**Comment by nguerrera** [comment]
Waiting for approval for Azure/cadl-azure#1304 before merging this as this will break cadl->cadl-azure flow if they're not merged together.

File: packages/compiler/core/module-resolver.ts
Comment (timotheeguerin) [inline packages/compiler/core/module-resolver.ts:89]: This was quite a bug!

**Review by timotheeguerin — APPROVED** [review]
Sweet, that's another thing I wanted to enable in eslint, we already had a bug a while back due to non-awaited promises, so easy to do.

**Pull request merged**
